crates/router/src/core/payment_methods/cards.rs

```diff
@@ -125,73 +135,156 @@ pub async fn update_customer_payment_method(
-            Err(err) => Err(report!(errors::VaultError::UnexpectedResponseError(
-                err.response
-            ))),
-        }?;
-        response
-    } else {
-        let card_id = generate_id(consts::ID_LENGTH, "card");
-        mock_add_card(db, &card_id, &card, None, None, Some(&customer_id)).await?
-    };
-
-    if let Some(false) = response.duplicate {
-        create_payment_method(db, &req, &customer_id, &response.card_id, merchant_id)
-            .await
-            .change_context(errors::VaultError::PaymentMethodCreationFailed)?;
-    } else {
-        match db.find_payment_method(&response.card_id).await {
-            Ok(_) => (),
-            Err(err) => {
-                if err.current_context().is_db_not_found() {
-                    create_payment_method(db, &req, &customer_id, &response.card_id, merchant_id)
-                        .await
-                        .change_context(errors::VaultError::PaymentMethodCreationFailed)?;
-                } else {
-                    Err(errors::VaultError::PaymentMethodCreationFailed)?;
-                }
-            }
+                .parse_struct("JweBody")
+                .change_context(errors::VaultError::ResponseDeserializationFailed)?;
+            let decrypted_payload =
+                payment_methods::get_decrypted_response_payload(jwekey, jwe_body)
+                    .await
+                    .change_context(errors::VaultError::SaveCardFailed)
+                    .attach_printable("Error getting decrypted response payload")?;
+            let stored_card_resp: payment_methods::StoreCardResp = decrypted_payload
+                .parse_struct("StoreCardResp")
+                .change_context(errors::VaultError::ResponseDeserializationFailed)?;
+            Ok(stored_card_resp)
         }
-    }
-    let payment_method_resp =
-        payment_methods::mk_add_card_response(card, response, req, merchant_id);
+        Err(err) => Err(report!(errors::VaultError::UnexpectedResponseError(
+            err.response
+        ))),
+    }?;
+    let store_card_payload = stored_card_response
+        .payload
+        .get_required_value("StoreCardRespPayload")
+        .change_context(errors::VaultError::SaveCardFailed)?;
+
+    create_payment_method(
+        db,
+        &req,
+        &customer_id,
+        &store_card_payload.card_reference,
+        merchant_id,
+    )
+    .await
+    .change_context(errors::VaultError::PaymentMethodCreationFailed)?;
+    let payment_method_resp = payment_methods::mk_add_card_response_hs(
+        card,
+        store_card_payload.card_reference,
+        req,
+        merchant_id,
+    );
     Ok(payment_method_resp)
 }
 
+#[instrument(skip_all)]
+pub async fn get_card_from_hs_locker<'a>(
+    state: &'a routes::AppState,
+    customer_id: &str,
+    merchant_id: &str,
+    card_reference: &'a str,
+) -> errors::RouterResult<payment_methods::Card> {
```

Comment:
Can we have the error, be something other than, `ApiErrorResponse`, as having errors like, VaultError, ApiClientError, and ParsingError in the current context. I suggest using some other error variant while returning as we can later convert it to ApiErrorResponse in the parent function